usr/lib/api/statistics.c

```diff
@@ -0,0 +1,265 @@
+
+    statistics->shm_handle = shm_open(statistics->shm_name, O_RDWR,
+                                      S_IRUSR | S_IWUSR);
+    if (statistics->shm_handle == -1) {
+        if (create) {
+            /* try to create it */
+            statistics->shm_handle = shm_open(statistics->shm_name,
+                                              O_CREAT | O_RDWR,
+                                              S_IRUSR | S_IWUSR);
+            if (statistics->shm_handle == -1) {
+                err = errno;
+                TRACE_ERROR("Failed to create SHM '%s': %s\n",
+                            statistics->shm_name,  strerror(err));
+                OCK_SYSLOG(LOG_ERR, "Failed to create SHM '%s': %s\n",
+                           statistics->shm_name, strerror(err));
+                return CKR_FUNCTION_FAILED;
+            }
+
+            if (fchmod(statistics->shm_handle, S_IRUSR | S_IWUSR) == -1) {
+                err = -errno;
+                TRACE_ERROR("Failed to change mode of SHM '%s': %s\n",
+                            statistics->shm_name,  strerror(err));
+                OCK_SYSLOG(LOG_ERR, "Failed to change mode of SHM '%s': %s\n",
+                           statistics->shm_name, strerror(err));
+                close(statistics->shm_handle);
+                shm_unlink(statistics->shm_name);
+                return CKR_FUNCTION_FAILED;
+            }
+            if (fchown(statistics->shm_handle, user, -1) != 0) {
+                err = -errno;
+                TRACE_ERROR("Failed to change owner of SHM '%s': %s\n",
+                            statistics->shm_name,  strerror(err));
+                OCK_SYSLOG(LOG_ERR, "Failed to change owner of SHM '%s': %s\n",
+                           statistics->shm_name, strerror(err));
+                close(statistics->shm_handle);
+                shm_unlink(statistics->shm_name);
+                return CKR_FUNCTION_FAILED;
+            }
+        } else {
+            err = errno;
+            TRACE_ERROR("Failed to open SHM '%s': %s\n",
+                        statistics->shm_name,  strerror(err));
+            OCK_SYSLOG(LOG_ERR, "Failed to open SHM '%s': %s\n",
+                       statistics->shm_name, strerror(err));
+            return CKR_FUNCTION_FAILED;
+        }
+    }
+
+    if (fstat(statistics->shm_handle, &stat_buf)) {
+        err = errno;
+        TRACE_ERROR("Failed to stat SHM '%s': %s\n",
+                    statistics->shm_name,  strerror(err));
+        OCK_SYSLOG(LOG_ERR, "Failed to stat SHM '%s': %s\n",
+                   statistics->shm_name, strerror(err));
+        close(statistics->shm_handle);
+        return CKR_FUNCTION_FAILED;
+    }
+
+    /*
+     * If the shared memory segment does not belong to the pkcs11 group or does
+     * not have correct permissions, do not use it.
+     */
+    if (stat_buf.st_uid != (user == -1 ? geteuid() : (uid_t)user) ||
+        stat_buf.st_gid != (user == -1 ? geteuid() : (uid_t)user) ||
```

Comment:
Comparing group id against user id here.  Should this be the id of the pkcs11 group?  Also I do not see the code that chowns the newly created segment to the pkcs11 group.

Reply:
oups,... . no pkcs11 group used here intentionally. the segment is supposed to belong only to the user.
Will change to only check st_uid and st_mode, but not st_gid